Nil pointer dereference when using IstioControlPlane to patch null properties

**Bug description**

I came across this bug when trying to customize an Istio installation using `IstioControlPlane` as described [here](https://istio.io/docs/reference/config/istio.operator.v1alpha12.pb/). I was trying to add EKS annotations to the `istio-ingressgateway` `Service` resource as follows:

```yaml
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  profile: default
  gateways:
    components:
      ingress_gateway:
        k8s:
          overlays:
          - apiVersion: v1
            kind: Service
            name: istio-ingressgateway
            patches:
            - path: metadata.annotations
              value:
                service.beta.kubernetes.io/aws-load-balancer-type: nlb
                service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0
```

By default, the `istio-ingressgateway` service looks like:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: istio-ingressgateway
  namespace: istio-system
  annotations:
  labels:
    app: istio-ingressgateway
    release: istio
    istio: ingressgateway
spec:
...
```

Which means that the `annotations` property holds a nil value. When you run `istioctl manifest generate` using the `IstioControlPlane` specification above you get:

```bash
> istioctl manifest generate -f istio/installation.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x2796e22]

goroutine 1 [running]:
istio.io/operator/pkg/util.IsSlice(0x0, 0x0, 0xc00081b1d0)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/util/reflect.go:47 +0x42
istio.io/operator/pkg/tpath.getPathContext(0xc00079fd70, 0xc000643ec0, 0x2, 0x2, 0xc000643ed0, 0x1, 0x1, 0xc000643e00, 0x20, 0x20, ...)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/tpath/tpath.go:163 +0xf15
istio.io/operator/pkg/tpath.getPathContext(0xc00079fc80, 0xc000643ec0, 0x2, 0x2, 0xc000643ec0, 0x2, 0x2, 0x0, 0x29, 0xc00081b508, ...)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/tpath/tpath.go:167 +0x100e
istio.io/operator/pkg/tpath.GetPathContext(...)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/tpath/tpath.go:65
istio.io/operator/pkg/patch.applyPatches(0xc00082ec00, 0xc00000ec10, 0x1, 0x1, 0xc000364a68, 0xc000843b80, 0x0, 0x1, 0xc0008d6fd0, 0x3)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/patch/patch.go:192 +0x31d
istio.io/operator/pkg/patch.YAMLManifestPatch(0xc000870f00, 0x2011, 0xc00077b670, 0xc, 0xc00000ec00, 0x1, 0x1, 0x0, 0x0, 0xc000870f00, ...)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/patch/patch.go:138 +0x3e0
istio.io/operator/pkg/component/component.renderManifest(0xc000a1d230, 0x30, 0x311f360, 0xffffffffffffff01, 0xc000a39050)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/component/component/component.go:784 +0x6dc
istio.io/operator/pkg/component/component.(*IngressComponent).RenderManifest(0xc000766538, 0xc000a38ff0, 0x100e8a3, 0x4cf1150, 0xc000a35b00)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/component/component/component.go:442 +0x3a
istio.io/operator/pkg/component/feature.renderComponents(0xc000862f40, 0x2, 0x2, 0x0, 0x7000102, 0x0, 0xffffffffffffffff)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/component/feature/feature.go:360 +0xc3
istio.io/operator/pkg/component/feature.(*GatewayFeature).RenderManifest(0xc000a1d2c0, 0x0, 0x0, 0x0, 0x0)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/component/feature/feature.go:262 +0x41
istio.io/operator/pkg/component/controlplane.(*IstioControlPlane).RenderManifest(0xc0008631e0, 0x0, 0x0, 0x0, 0xc0004d4d00)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/pkg/component/controlplane/control_plane.go:69 +0xcd
istio.io/operator/cmd/mesh.genManifests(0x7ffeefbff47f, 0x17, 0x0, 0x0, 0xc000975d00, 0xc000b81d48, 0x0, 0x0, 0x0)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/cmd/mesh/manifest-common.go:119 +0x252
istio.io/operator/cmd/mesh.manifestGenerate(0xc00086610c, 0xc0001c2b00, 0xc00081bd48)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/cmd/mesh/manifest-generate.go:75 +0x166
istio.io/operator/cmd/mesh.manifestGenerateCmd.func2(0xc0001aef00, 0xc0006bb440, 0x0, 0x2)
	istio.io/operator@v0.0.0-20191202210744-2689e5c3357f/cmd/mesh/manifest-generate.go:60 +0x100
github.com/spf13/cobra.(*Command).execute(0xc0001aef00, 0xc0006bb3c0, 0x2, 0x2, 0xc0001aef00, 0xc0006bb3c0)
	github.com/spf13/cobra@v0.0.5/command.go:830 +0x2aa
github.com/spf13/cobra.(*Command).ExecuteC(0xc00073ea00, 0x4, 0x4, 0xc00073ea00)
	github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	istio.io/istio@/istioctl/cmd/istioctl/main.go:29 +0x76
```

[This commit](https://github.com/ashlineldridge/operator/commit/0acb903e0f10f93102744d7a1b77e873b3019ac1) demonstrates the issue and a possible fix. Happy to submit a PR.

**Expected behavior**

It should be possible to patch YAML properties whose value is null.

**Steps to reproduce the bug**

As described above.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

Istio version:
```
client version: 1.4.1
control plane version: 1.4.1
```

Kubectl version:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-14T04:24:34Z", GoVersion:"go1.12.13", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.8-eks-b7174d", GitCommit:"b7174db5ee0e30c94a0b9899c20ac980c0850fc8", GitTreeState:"clean", BuildDate:"2019-10-18T17:56:01Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Downloaded from the GitHub releases page.

**Environment where bug was observed (cloud vendor, OS, etc)**
On macOS 10.14.6.

--- a/pkg/patch/patch.go
+++ b/pkg/patch/patch.go
@@ -40,7 +40,7 @@ MODIFY
 1. set v1 to v1new
 
   path: a.b.[name:n1].value
-  value: v1
+  value: v1new
 
 2. set vv1 to vv3
 

--- a/pkg/patch/patch_test.go
+++ b/pkg/patch/patch_test.go
@@ -38,6 +38,7 @@ a:
     - v1
     - v2
     - v3_regex
+  c:
 `
 	tests := []struct {
 		desc    string
@@ -65,6 +66,7 @@ a:
     - v2
     - v3_regex
     name: n2
+  c:
 `,
 		},
 		{
@@ -86,6 +88,7 @@ a:
     - v2
     - v3_regex
     name: n2
+  c:
 `,
 		},
 		{
@@ -107,6 +110,7 @@ a:
     - v3
     - v3_regex
     name: n2
+  c:
 `,
 		},
 		{
@@ -125,6 +129,7 @@ a:
     - v2
     - v3_regex
     name: n2
+  c:
 `,
 		},
 		{
@@ -144,6 +149,7 @@ a:
     - v1
     - v3_regex
     name: n2
+  c:
 `,
 		},
 		{
@@ -163,8 +169,34 @@ a:
     - v1
     - v2
     name: n2
+  c:
 `,
-		}}
+		},
+		{
+			desc: "UpdateNullNode",
+			path: `a.c`,
+			value: `
+      d: n3`,
+			want: `
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: istio-citadel
+  namespace: istio-system
+a:
+  b:
+  - name: n1
+    value: v1
+  - name: n2
+    list: 
+    - v1
+    - v2
+    - v3_regex
+  c:
+    d: n3
+`,
+		},
+	}
 	for _, tt := range tests {
 		t.Run(tt.desc, func(t *testing.T) {
 			rc := &v1alpha2.KubernetesResourcesSpec{}
@@ -355,7 +387,7 @@ spec:
 `,
 		},
 		{
-			desc: "DeleteLeafListentry",
+			desc: "DeleteLeafListEntry",
 			path: `spec.template.spec.containers.[name:galley].command.[--validation-webhook-config-file]`,
 			want: `
 apiVersion: extensions/v1beta1

--- a/pkg/util/reflect.go
+++ b/pkg/util/reflect.go
@@ -21,19 +21,28 @@ import (
 	"github.com/kr/pretty"
 )
 
+// kindOf returns the reflection Kind that represents the dynamic type of value.
+// If value is a nil interface value, kindOf returns reflect.Invalid.
+func kindOf(value interface{}) reflect.Kind {
+	if value == nil {
+		return reflect.Invalid
+	}
+	return reflect.TypeOf(value).Kind()
+}
+
 // IsString reports whether value is a string type.
 func IsString(value interface{}) bool {
-	return reflect.TypeOf(value).Kind() == reflect.String
+	return kindOf(value) == reflect.String
 }
 
 // IsPtr reports whether value is a ptr type.
 func IsPtr(value interface{}) bool {
-	return reflect.TypeOf(value).Kind() == reflect.Ptr
+	return kindOf(value) == reflect.Ptr
 }
 
 // IsMap reports whether value is a map type.
 func IsMap(value interface{}) bool {
-	return reflect.TypeOf(value).Kind() == reflect.Map
+	return kindOf(value) == reflect.Map
 }
 
 // IsMapPtr reports whether v is a map ptr type.
@@ -44,17 +53,16 @@ func IsMapPtr(v interface{}) bool {
 
 // IsSlice reports whether value is a slice type.
 func IsSlice(value interface{}) bool {
-	return reflect.TypeOf(value).Kind() == reflect.Slice
+	return kindOf(value) == reflect.Slice
 }
 
 func IsStruct(value interface{}) bool {
-	return reflect.TypeOf(value).Kind() == reflect.Struct
+	return kindOf(value) == reflect.Struct
 }
 
 // IsSlicePtr reports whether v is a slice ptr type.
 func IsSlicePtr(v interface{}) bool {
-	t := reflect.TypeOf(v)
-	return t.Kind() == reflect.Ptr && t.Elem().Kind() == reflect.Slice
+	return kindOf(v) == reflect.Ptr && reflect.TypeOf(v).Elem().Kind() == reflect.Slice
 }
 
 // IsSliceInterfacePtr reports whether v is a slice ptr type.
@@ -131,7 +139,7 @@ func IsValueNil(value interface{}) bool {
 	if value == nil {
 		return true
 	}
-	switch reflect.TypeOf(value).Kind() {
+	switch kindOf(value) {
 	case reflect.Slice, reflect.Ptr, reflect.Map:
 		return reflect.ValueOf(value).IsNil()
 	}
@@ -206,7 +214,7 @@ func IsEmptyString(value interface{}) bool {
 	if value == nil {
 		return true
 	}
-	switch reflect.TypeOf(value).Kind() {
+	switch kindOf(value) {
 	case reflect.String:
 		if _, ok := value.(string); ok {
 			return value.(string) == ""

